fetch shader

trenouf opened the pull request:
The top five commits implement a fetch shader to enable compilation of a VS without vertex input info:

c47bc92e2 Generate fetch shader in linker if VS is fetchless
255e8f225 Fetchless vertex shader
d7711028e New vertex fetch pass
cdfde289f Factored out addFunctionArgs
43e4e8b7e Put AddressExtender in its own header file

trenouf commented:
V4: Fixed PipelineVsFs_RelocConst test to have vertex input state.

amdvlk-admin commented:
### Test summary for commit 175682be89b99f5d956f14240afc3c59f8ffd528
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 85170384d0ee5447ac267d478f7a21bc036ea71c</li>
<li>SPVGEN: dev 704f4adddb6867d633431c2c9c2c0bfa6d294ce2</li>
<li>XGL: dev 9c4fe791366e730131a0d99a7519855068f216ef</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 3a1103f746f4d782271bd32d67274d6260c5dc7f</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 42435/42436 (100.0%)</li>
<li><details><summary>Failed: 1/42436 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.spirv_assembly.type.vec3.i16.sub_test_high_part_zero_geom
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>


</ul></details>

amdvlk-admin commented:
### Test summary for commit dbe81ba5f8b456d472519f14c6f6f0adfcee648b
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 85170384d0ee5447ac267d478f7a21bc036ea71c</li>
<li>SPVGEN: dev 704f4adddb6867d633431c2c9c2c0bfa6d294ce2</li>
<li>XGL: dev 9c4fe791366e730131a0d99a7519855068f216ef</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 3a1103f746f4d782271bd32d67274d6260c5dc7f</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

amdrexu commented on lgc/patch/NggPrimShader.cpp:
```diff
@@ -222,6 +224,21 @@ FunctionType *NggPrimShader::generatePrimShaderEntryPointType(uint64_t *inRegMas
+    unsigned vertexFetchCount = m_pipelineState->getPalMetadata()->getVertexFetchCount();
+    if (vertexFetchCount != 0) {
+      // TODO: This probably will not work with culling.
+      if (m_hasGs)
```
I think we only have to report error in this case for NGG: VS+FS (with culling). When we have GS, culling could be done regardless of whether you separate fetch shader from VS because the culling is based on GS output primitive attribute data (not VS fetch data).

nhaehnle commented on lgc/include/lgc/util/AddressExtender.h:
```diff
@@ -0,0 +1,97 @@
+  // @param ptrTy : Type to cast pointer to
+  // @param builder : IRBuilder to use, already set to the required insert point
+  // @return : 64-bit pointer value
+  llvm::Instruction *extend(llvm::Value *addr32, unsigned highHalf, llvm::Type *ptrTy, llvm::IRBuilder<> &builder) {
+    llvm::Value *ptr = nullptr;
+    if (highHalf == HighAddrPc) {
+      // Extend with PC.
+      ptr = builder.CreateInsertElement(getPc(), addr32, uint64_t(0));
+    } else {
+      // Extend with given value
+      ptr = builder.CreateInsertElement(llvm::UndefValue::get(llvm::VectorType::get(builder.getInt32Ty(), 2)), addr32,
+                                        uint64_t(0));
+      ptr = builder.CreateInsertElement(ptr, builder.getInt32(highHalf), 1);
+    }
+    ptr = builder.CreateBitCast(ptr, builder.getInt64Ty());
+    return llvm::cast<llvm::Instruction>(builder.CreateIntToPtr(ptr, ptrTy));
+  }
```
The implementation should really go into a .cpp file.

nhaehnle commented on lgc/state/ShaderStage.cpp:
```diff
@@ -96,3 +98,111 @@ const char *lgc::getShaderStageAbbreviation(ShaderStage shaderStage) {
+  }
+
+  // Copy attributes and shader stage from the old function.
+  newFunc->setAttributes(oldFunc->getAttributes());
+  setShaderStage(newFunc, getShaderStage(oldFunc));
+
+  // Replace uses of the old args.
+  // Set inreg attributes correctly. We have to use removeAttr because arg attributes are actually attached
+  // to the Function, and the attribute copy above copied the arg attributes at their original arg numbers.
+  // Also set name of each new arg that comes from old arg.
+  for (unsigned idx = 0; idx != argTys.size(); ++idx) {
+    Argument *arg = newFunc->getArg(idx);
+    if ((inRegMask >> idx) & 1)
+      arg->addAttr(Attribute::InReg);
+    else if (!oldFuncTy->params().empty())
+      arg->removeAttr(Attribute::AttrKind::InReg);
+  }
+  for (unsigned idx = 0; idx != oldFuncTy->params().size(); ++idx) {
+    Argument *arg = newFunc->getArg(idx + argTys.size());
+    Argument *oldArg = oldFunc->getArg(idx);
+    arg->setName(oldArg->getName());
+    oldArg->replaceAllUsesWith(arg);
+    if (oldArg->hasInRegAttr())
+      arg->addAttr(Attribute::InReg);
+    else
+      arg->removeAttr(Attribute::AttrKind::InReg);
+  }
```
It should be possible to do this more idiomatically and efficiently, by using `AttributeList::get` once, passing it:
* the original function attributes
* the original return attributes
* a vector that contains AttributeSets for all the new arguments as well as the shifted old arguments.

nhaehnle commented on lgc/patch/VertexFetch.cpp:
```diff
@@ -25,28 +25,140 @@
+    Constant *int8;     // < 0, 0, 0, 1 >
+    Constant *int16;    // < 0, 0, 0, 1 >
+    Constant *int32;    // < 0, 0, 0, 1 >
+    Constant *int64;    // < 0, 0, 0, 0, 0, 0, 0, 1 >
```
Shouldn't this be `< 0, 0, 0, 0, 0, 0, 1, 0 >`? Though I suppose that's not your bug...

amdrexu replied:
Yes, should be what you have said. It is my bug :)

trenouf commented:
> I don't look through this PR. I will do this in following days. I have a top-level question: do we fall back to full pipeline compilation when we determine we couldn't generate separate fetch shader? If yes, the change is currently safe to NGG culling. If no, maybe more testing is needed to verify the correctness with culling enabled.

That is the plan, to fall back to full pipeline compilation if the link fails for that kind of reason. However, that is not implemented yet. See the TODO towards the end of linkRelocatableShaderElf in llpcCompiler.cpp.

Also, Steven has a plan for spotting things like that ahead of time if possible.

trenouf commented:
> That is the plan, to fall back to full pipeline compilation if the link fails for that kind of reason.

Or if the shader compile fails (as it will in this case, for culling NGG).

trenouf commented:
V2: Put code in a .cpp file.
    V2: Improved handling of attributes.
    V5: Changed failure condition to non-GS culling.
    V5: Adjusted lgc test for abandoning attempt to stop LLVM adding
        s_end_code padding.

amdvlk-admin commented:
### Test summary for commit dde2ce42a97fc4f9811311e2d891f0d96a8a7c29
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 85170384d0ee5447ac267d478f7a21bc036ea71c</li>
<li>SPVGEN: dev 704f4adddb6867d633431c2c9c2c0bfa6d294ce2</li>
<li>XGL: dev 9c4fe791366e730131a0d99a7519855068f216ef</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 3a1103f746f4d782271bd32d67274d6260c5dc7f</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 42724/42725 (100.0%)</li>
<li><details><summary>Failed: 1/42725 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.spirv_assembly.type.vec3.u16.slessthanequal_tesse
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>


</ul></details>

amdrexu commented on lgc/elfLinker/GlueShader.cpp:
```diff
@@ -0,0 +1,284 @@
+          replacement = builder.CreateBitCast(
+              fetchFunc->getArg(m_vsEntryRegInfo.sgprCount + m_vsEntryRegInfo.instanceId), builder.getInt32Ty());
+          break;
+        default:
```
It seems DrawIndex is not listed here?

trenouf replied:
I think a vertex fetch never uses DrawIndex.

amdrexu replied:
OK. I didn't notice this is for vertex fetching only.

amdrexu commented on lgc/include/lgc/state/AbiMetadata.h:
```diff
@@ -207,11 +207,49 @@ constexpr unsigned int mmSPI_SHADER_USER_DATA_HS_0 = 0x2D0C; // For GFX9+, Used
+};
+
+// General RSRC2 register
+union SPI_SHADER_PGM_RSRC2 {
```
Which fields are used actually? Those listed? LDS_SIZE is not always present for each shader stage though previous field definitions, like SCRATCH_EN, USER_SGPR, etc., are identical. Also, the bitwidth of this field is not always 9 bits. GFX10 HS and GS have 10 bits while pre-GFX10 only have 9 bits. Furthermore, the field offset is not the same:)

trenouf replied:
Good point. I'm only using USER_SGPR, so I will remove the other fields.

amdrexu commented on lgc/include/lgc/state/AbiMetadata.h:
```diff
@@ -242,4 +280,20 @@ union PA_CL_CLIP_CNTL {
+    unsigned int HS_W32_EN : 1;
+    unsigned int GS_W32_EN : 1;
+    unsigned int VS_W32_EN : 1;
+    unsigned int NGG_WAVE_ID_EN : 1;
```
I think here we need the field PRIMGEN_EN for NGG, not NGG_WAVE_ID_EN.

trenouf replied:
Again, I will remove the ones I am not using.

amdrexu commented on lgc/state/ShaderStage.cpp:
```diff
@@ -96,3 +96,80 @@ const char *lgc::getShaderStageAbbreviation(ShaderStage shaderStage) {
 }
+
+// =====================================================================================================================
+// Add args to a function. The new args are put before any existing ones. This creates a new function with the
```
So this function might finally leave a empty old function because referenced blocks are moved to the new one?

trenouf replied:
Yes; the caller removes it. In future changes, the function being mutated might be a subfunction, and it is up to the caller of addFunctionArgs to change uses of the old function to the new function, which must be done before erasing the old function.

amdrexu commented:
> That is the plan, to fall back to full pipeline compilation if the link fails for that kind of reason. However, that is not implemented yet. See the TODO towards the end of linkRelocatableShaderElf in llpcCompiler.cpp.

I see. Do we still have full pipeline compilation as the default mode or can we force it with panel option? I ask this because if fetchless vertex is the default mode, we are sure to run into problems for NGG VS culling. Once the PR is submitted and follow-up TODO change is not finished in a short time, game quality testing and performance testing of NGG culling will be broken immediately.

trenouf commented:
> > That is the plan, to fall back to full pipeline compilation if the link fails for that kind of reason. However, that is not implemented yet. See the TODO towards the end of linkRelocatableShaderElf in llpcCompiler.cpp.
> 
> I see. Do we still have full pipeline compilation as the default mode or can we force it with panel option? I ask this because if fetchless vertex is the default mode, we are sure to run into problems for NGG VS culling. Once the PR is submitted and follow-up TODO change is not finished in a short time, game quality testing and performance testing of NGG culling will be broken immediately.

The default is still full pipeline compilation. Steven has an option to enable shader compilation and linking, so he can test it, but it is not default.

amdrexu commented:
> Steven has an option to enable shader compilation and linking, so he can test it, but it is not default.

That's good. We can have smooth transition.

trenouf commented:
V6: Only define bits we use in RSRC2 and VGT_SHADER_STAGES_EN.

trenouf commented:
Rebased.

amdvlk-admin commented:
### Test summary for commit ff6f88ce892aa87eeb41ecf654fffe7f64cfa25e
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev 82390674c40402c81dbf311a5ee488fb972894f5</li>
<li>SPVGEN: dev 2f679769a8491f423f5cf2f06adfa2870a82935d</li>
<li>XGL: dev 116d28a446eba3a9f8e6f577bab34fc07e090993</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev e90e9d404b64f21a69c23e2a8709e4eb56ba04e4</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

trenouf commented:
Ping for approval.

trenouf commented:
Rebased.

trenouf commented:
Jenkins Pipeline passed all three CTS runs, then went wrong reporting the results.
